### PR TITLE
feat: add uniswap VIP support

### PR DESCRIPTION
**Pull request by dorothy-zbornak**
Also add `includedSources` query param support to end points. I hate how much work this totally uncalled for feature took. :weary: 

# Description

# Testing Instructions

# Checklist

-   [ ] Update [documentation](https://github.com/0xProject/website/blob/development/mdx/api/index.mdx) as needed. **Website Documentation PR:**
-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -422,6 +423,9 @@ export const GAS_SCHEDULE_V1: FeeSchedule = {
     [ERC20BridgeSource.MultiHop]: fillData => {
         const firstHop = (fillData as MultiHopFillData).firstHopSource;
         const secondHop = (fillData as MultiHopFillData).secondHopSource;
+        if (!firstHop || !secondHop) {
```
Comment (dorothy-zbornak) [inline src/config.ts:426]: This code path will get called even if `MultiHop` samples nothing now (ie, it's disabled), so we need this catch.

File: src/config.ts [hunk 2]
```diff
@@ -350,6 +350,7 @@ export const GAS_SCHEDULE_V0: FeeSchedule = {
     [ERC20BridgeSource.Bancor]: () => 4.5e5,
     [ERC20BridgeSource.MStable]: () => 8.5e5,
     [ERC20BridgeSource.Mooniswap]: () => 3.5e5,
+    [ERC20BridgeSource.MultiHop]: () => 0,
```
Comment (dorothy-zbornak) [inline src/config.ts:353]: This code path will get called even if `MultiHop` samples nothing now (ie, it's disabled), so we need this stub.

File: src/utils/parse_utils.ts [hunk 3]
```diff
@@ -41,85 +44,82 @@ export const parseUtils = {
...
             return {
-                excludedSources: parseUtils.parseStringArrForERC20BridgeSources(request.excludedSources.split(',')),
+                excludedSources: parseUtils.parseStringArrForERC20BridgeSources(excludedIds),
+                includedSources: [],
                 nativeExclusivelyRFQT: false,
             };
         }
 
-        if (request.includedSources !== undefined) {
-            // Only RFQT is eligible as of now
```
Comment (dorothy-zbornak) [inline src/utils/parse_utils.ts:53]: So now `RFQT` can coexist with other sources in `includedSources`.

**Review by dekz — APPROVED** [review]

File: src/handlers/swap_handlers.ts [hunk 4]
```diff
@@ -183,6 +183,10 @@ export class SwapHandlers {
                 req.query.excludedSources === undefined
                     ? []
                     : parseUtils.parseStringArrForERC20BridgeSources((req.query.excludedSources as string).split(',')),
+            includedSources:
+                req.query.includedSources === undefined
+                    ? []
+                    : parseUtils.parseStringArrForERC20BridgeSources((req.query.includedSources as string).split(',')),
```
Comment (PirosB3) [inline src/handlers/swap_handlers.ts:189]: `includedSources` can only be `RFQT`. Doesn't this added code mean that it could be any ERC20 source?

Comment (dorothy-zbornak) [reply src/handlers/swap_handlers.ts:189]: Yes. We now support an `includedSources` option in AS.

Comment (dorothy-zbornak) [reply src/handlers/swap_handlers.ts:189]: This change should still allow you to ask for RFQT only.

File: src/utils/parse_utils.ts [hunk 5]
```diff
@@ -41,85 +44,82 @@ export const parseUtils = {
...
+                        field: 'intentOnFilling',
+                        code: ValidationErrorCodes.RequiredField,
+                        reason: ValidationErrorReasons.RequiresIntentOnFilling,
                     },
                 ]);
             }
+
+            return {
+                excludedSources: [],
+                includedSources: parseUtils.parseStringArrForERC20BridgeSources(['0x', ...includedIds]),
```
Comment (PirosB3) [inline src/utils/parse_utils.ts:103]: `RFQT` is not part of ERC20Source, will it get removed when `parseStringArrForERC20BridgeSources` is called?

File: src/utils/parse_utils.ts [hunk 6]
```diff
@@ -41,85 +44,82 @@ export const parseUtils = {
...
+                nativeExclusivelyRFQT: !includedIds.includes('0x') && !includedIds.includes('Native'),
```
Comment (PirosB3) [inline src/utils/parse_utils.ts:104]: Could we simply make RFQT a special edge case here? for example, as of today there is no taker that requests exclusive RFQT liquidity alongside another on-chain source. If we feel good about making this assumption, we can add the following checks and statements in the code:

- If `RFQT` is added to `includedSources`, then `includedIds.length` must == 1
- `nativeExclusivelyRFQT: includedIds.includes("RFQT")`. So you could keep `L104` as `nativeExclusivelyRFQT: true`.

File: src/utils/parse_utils.ts [hunk 7]
```diff
@@ -24,9 +24,12 @@ export const parseUtils = {
         request: ParseRequestForExcludedSourcesParams,
         validApiKeys: string[],
         endpoint: 'price' | 'quote',
-    ): { excludedSources: ERC20BridgeSource[]; nativeExclusivelyRFQT: boolean } {
+    ): { excludedSources: ERC20BridgeSource[]; includedSources: ERC20BridgeSource[]; nativeExclusivelyRFQT: boolean } {
```
Comment (PirosB3) [inline src/utils/parse_utils.ts:27]: This return type has grow a little, have you considered making it it's own type?

File: src/handlers/meta_transaction_handlers.ts [hunk 8]
```diff
@@ -369,6 +373,10 @@ const parseGetTransactionRequestParams = (req: express.Request): GetTransactionR
         req.query.excludedSources === undefined
             ? undefined
             : parseUtils.parseStringArrForERC20BridgeSources((req.query.excludedSources as string).split(','));
+    const includedSources =
```
Comment (PirosB3) [inline src/handlers/meta_transaction_handlers.ts:376]: I'm really happy that you added this! thanks!

**Review by PirosB3 — CHANGES_REQUESTED** [review]
Sorry for the delay! I ran a first review :)

File: src/utils/parse_utils.ts (hunk 5 above)
Comment (dorothy-zbornak) [reply src/utils/parse_utils.ts:103]: Yes.

File: src/utils/parse_utils.ts (hunk 6 above)
Comment (dorothy-zbornak) [reply src/utils/parse_utils.ts:104]: Idk, I could see simbot playing with a mix of sources and excluding open orderbook. The behavior for a taker passing in just `includedSources=RFQT` is still the same, right? What is the motivation for not allowing other included sources to coexist with RFQT?

**Comment by dekz** [comment]
deploy staging

**Comment by dekz** [comment]
deploy production

**Pull request merged**